### PR TITLE
KAFKA-3502: KStreamTestDriver needs to be closed after the test case

### PR DESCRIPTION
**Pull request by guozhangwang**
Found a few recently added unit tests did not close KStreamTestDriver after the test itself is closed; this can cause RocksDB virtual function called if the contained topology has some persistent store since they will be initialized but not closed in time.

MINOR fix: found that when closing KStreamTestDriver, we need to first flushing all stores before closing any of them; this is triggered from the `KTableKTableLeftJoin.shouldNotThrowIllegalStateExceptionWhenMultiCacheEvictions`.

MINOR fix: in CachingXXXStore, the `name` field is actually used as the cache's namespace, not really the store name or its corresponding topic name. Fixed it by renaming it to `cacheName` and use `this.name()` elsewhere which will call the underlying store's name.

### CODE REVIEW

**Comment by guozhangwang** [comment]
Ping @dguy @mjsax @enothereska for reviews.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java
Comment (ewencp) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java:70]: I only checked a few, but it looks like almost all tests use these basically all have the same cleanup and clearly it's easy to forget to close it. Perhaps a shared, albeit very simple, base class would make it less likely that people forget to close the driver in tests?

Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java:70]: We are aware of that and are planning to do it in a separate JIRA: https://issues.apache.org/jira/browse/KAFKA-3623

**Review by ewencp — APPROVED** [review]
@guozhangwang Follow up refactoring makes sense. Otherwise this LGTM and seems straightforward, but I'll leave it up to you whether to merge or wait for additional reviews.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/GlobalKTableJoinsTest.java
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/GlobalKTableJoinsTest.java:76]: This is redundant. Every test gets a new instance of the class.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java
Comment (dguy) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java:70]: Personally not a fan of introducing base classes for testing. IMO it would be better to change `KStreamTestDriver` to extend `org.junit.rules.ExternalResource`. It can override the `after` method and call close there.

**Review by dguy — CHANGES_REQUESTED** [review]

**Comment by enothereska** [comment]
LGTM

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/GlobalKTableJoinsTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/GlobalKTableJoinsTest.java:76]: This is to prevent some test cases that do not actually create a new `driver` instance: in some test classes we have only a few test case that create a new `driver`.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java:70]: I agree, and personally I also prefer the `ExternalResource` way and that is as I said in KAFKA-3623. But we can still debate over Ewen's suggestion v.s. `ExternalResource` since not all tests creates an instance of `driver` which makes it a bit tricky.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/GlobalKTableJoinsTest.java
Comment (dguy) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/GlobalKTableJoinsTest.java:76]: In this test though it is always created, right?

**Comment by guozhangwang** [comment]
@dguy Replied to your comments. But since you opted to "request changes" I'd try to confirm if you are +1 for merging it or not.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java
Comment (dguy) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java:70]: It is fine if the external resouce is `null`. junit won't try and invoke a method on a `null` external resource.

**Review by dguy — APPROVED** [review]
LGTM

**Comment by guozhangwang** [comment]
Merged to trunk and 0.10.2.

**Pull request closed without merging**
